src/certgen_certificate_generator_server.cc

```diff
@@ -327,22 +325,19 @@ namespace certgen
 
         fty::CertificateConfig config = loadConfig (certgenConfig.version(), certgenConfig.certConf());
 
-        std::string csrPem;
+        fty::Keys keyPair(generateKeys(certgenConfig.keyConf()));
 
-        // TODO should we implement copy ctor on Keys??
-        m_csrKey.reset(new Keys(generateKeys(certgenConfig.keyConf())));
+        fty::CsrX509 csr = CsrX509::generateCsr(keyPair, config);
 
-        try
-        {
-            CsrX509 csr = CsrX509::generateCsr(*(m_csrKey.get()), config);
-            csrPem = csr.getPem();
-        }
-        catch(std::runtime_error &e)
+        // replace existing csr for a given service, if any
```

Comment:
I could't find a better solution for this, there is a replace function for std::map [`insert_or_assign`](https://en.cppreference.com/w/cpp/container/map/insert_or_assign), but it's only available in std >= c++17 and gcc >= 5.x, I guess

Reply:
Good to me.